### PR TITLE
make shape inference of BatchNorm layout neutral

### PR DESCRIPTION
**Pull request by guopinglong**
The original shape inference logic is very specific to NCHW. This change is to make it a little be more layout neutral. So other layouts can be incorporated more easily.

### CODE REVIEW

**Comment by guopinglong** [comment]
Nice comment. Refactored to use the axis variable directly instead of allocating a new one (layout).

File: src/top/nn/nn.cc [hunk 1]
```diff
@@ -117,19 +117,33 @@ DMLC_REGISTER_PARAMETER(BatchNormParam);
...
       << "Input:[data, gamma, beta, moving_mean, moving_var]";
   CHECK_EQ(out_shape->size(), 3U);
   const TShape &dshape = in_shape->at(0);
   if (dshape.ndim() == 0) return false;
-  TShape bshape({dshape[1]});
-  NNVM_ASSIGN_INPUT_SHAPE(attrs, *in_shape, 1, bshape);
-  NNVM_ASSIGN_INPUT_SHAPE(attrs, *in_shape, 2, bshape);
-  NNVM_ASSIGN_INPUT_SHAPE(attrs, *in_shape, 3, bshape);
-  NNVM_ASSIGN_INPUT_SHAPE(attrs, *in_shape, 4, bshape);
+  auto bshape = [&](int axis) {
```
Comment (tqchen) [inline src/top/nn/nn.cc]: do we really need switch in here? would dshape[axis] suffice?

**Comment by guopinglong** [comment]
Yes. The change should be minor in this case.

**Comment by guopinglong** [comment]
Test case added.

**Pull request merged**